var/spack/repos/builtin/packages/hdf5/package.py

```diff
@@ -156,6 +156,9 @@ class Hdf5(CMakePackage):
         "+fortran", when="@1.13.3:^cmake@:3.22", msg="cmake_minimum_required is not set correctly."
     )
 
+    # See https://github.com/HDFGroup/hdf5/pull/4147
+    requires("^zlib-ng~new_strategies", when="@:1.14.3,develop-1.8:develop-1.12 ^zlib-ng")
+
```

Comment:
Can you try:
```
depends_on("zlib-ng~new_strategies", when="@:1.14.3,develop-1.8:develop-1.12 ^[virtuals=zlib-api] zlib-ng")
```
?

Reply:
It works that way, yes. I have to say though that it feels a bit like playing a lottery.

Reply:
:sweat: Yeah, I can understand that. Let's say that the way this is implemented right now, using `depends_on` or `requires` for this edge case gives the same set of possible solutions, but orders them differently (so the optimal one is different).

Reply:
did you ever agree on what to do about this?  Are we going with @alalazo's suggestion or @skosukhin's?

Reply:
I'm fine with following @alalazo's suggestion. I just need to find time to update the PR.

Reply:
@alalazo I guess we need some extra magic here to pass the gitlab-ci. Could you help with that, please?

Reply:
I assume that due to the changes in the PR, different dependents of `zlib-ng` expect the package in different configurations, which the Windows job doesn't like.

Reply:
For the record, and example of the error is https://gitlab.spack.io/spack/spack/-/jobs/12900489

![Screenshot from 2024-10-09 11-34-37](https://github.com/user-attachments/assets/b7cc4fdf-0a44-4a10-b3e6-e4bd4844e4de)